### ISSUE TITLE
Restart counter for manual restarts

### ISSUE THREAD

**Issue by martinvonwittich** [systemd/systemd#29348]
### Component

systemd

### Is your feature request related to a problem? Please describe

Since #4126, `systemd` has a restart counter `NRestarts` that counts how many times `systemd` had to automatically restart a service. This has proven to be a very useful troubleshooting aid for us - we've set up a login script on all our systems that warns when there are services with NRestarts > 0, which can help our supporters to quickly identify problems caused by crashing services.

Unfortunately, today we had an issue with a service caught in a restart loop that this warning didn't show. The reason was that a faulty shell script kept restarting Samba with `systemctl restart samba-ad-dc` in an endless loop:

```
host ~ # grep samba-ad-dc.service /var/log/syslog | tail -n 20   
Sep 27 16:30:18 iserv systemd[1]: samba-ad-dc.service: Main process exited, code=exited, status=127/n/a
Sep 27 16:30:18 iserv systemd[1]: samba-ad-dc.service: Failed with result 'exit-code'.
Sep 27 16:30:18 iserv systemd[1]: Stopped samba-ad-dc.service.
Sep 27 16:30:18 iserv systemd[1]: samba-ad-dc.service: Consumed 1min 37.353s CPU time.
Sep 27 16:30:18 iserv systemd[1]: Starting samba-ad-dc.service...
Sep 27 16:30:18 iserv systemd[1]: Started samba-ad-dc.service.
Sep 27 16:31:07 iserv systemd[1]: Stopping samba-ad-dc.service...
Sep 27 16:31:07 iserv systemd[1]: samba-ad-dc.service: Main process exited, code=exited, status=127/n/a
Sep 27 16:31:07 iserv systemd[1]: samba-ad-dc.service: Failed with result 'exit-code'.
Sep 27 16:31:07 iserv systemd[1]: Stopped samba-ad-dc.service.
Sep 27 16:31:07 iserv systemd[1]: samba-ad-dc.service: Consumed 23.191s CPU time.
Sep 27 16:31:07 iserv systemd[1]: Starting samba-ad-dc.service...
Sep 27 16:31:08 iserv systemd[1]: Started samba-ad-dc.service.
Sep 27 16:45:17 iserv systemd[1]: Stopping samba-ad-dc.service...
Sep 27 16:45:17 iserv systemd[1]: samba-ad-dc.service: Main process exited, code=exited, status=127/n/a
Sep 27 16:45:30 iserv systemd[1]: samba-ad-dc.service: Failed with result 'exit-code'.
Sep 27 16:45:30 iserv systemd[1]: Stopped samba-ad-dc.service.
Sep 27 16:45:30 iserv systemd[1]: samba-ad-dc.service: Consumed 1min 37.114s CPU time.
Sep 27 16:45:30 iserv systemd[1]: Starting samba-ad-dc.service...
Sep 27 16:45:31 iserv systemd[1]: Started samba-ad-dc.service.
```

```
host ~ # grep samba-ad-dc.service /var/log/syslog | grep 'Main process exited' | wc -l
442
```

`NRestarts` doesn't count these restarts:

```
host ~ # systemctl show -p NRestarts samba-ad-dc.service 
NRestarts=0
```

From what I can tell from #6495, this is an intentional limitation of `NRestarts` because a restart counter for manual restarts would need to be stored in the journal, and not in the in-memory unit.

### Describe the solution you'd like

An additional counter that complements `NRestarts`, and that can also count manual restarts from `systemctl restart`. As far as I understand the PR that implemented `NRestarts`, this would also need some way to store persistent unit metrics in the journal, but I don't know the `systemd` internals well enough to determine if this already exists, or if it would still need to be implemented.

### Describe alternatives you've considered

I believe the only workaround that is currently possible is to count the restarts by parsing the journal / `syslog`. This has the disadvantage that it's not necessarily accurate or fast - in pathological cases where the log is flooded with useless messages, the journal will probably expire log entries very quickly (which leads us to miss restart events), while the `syslog` would likely grow to enormous sizes (which would cause our log parsing to become very slow and IO-intensive). Parsing the log would probably be a bad idea in a login script, so I don't see a way how to quickly query this information to show it on login.

### The systemd version you checked that didn't have the feature you are asking for

247

**Comment by Werkov** [comment]
> a faulty shell script kept restarting Samba with systemctl restart samba-ad-dc in an endless loop

So that scripts should be fixed first, no? (Or the service unit file, configure the Restart*= directives to rely on automatic systemd restarts -> you get the counter info.)

Alternatively, you can protect against mistaken "controlling" entity by tweaking `StartLimitIntervalSec=` and `StartLimitBurst=` which will kick in and raise messages about unexepected behavior.

**Comment by martinvonwittich** [comment]
> So that scripts should be fixed first, no?

Yes, of course. But that's not what this ticket is about - the backstory with the faulty script was just meant to describe why a manual restart counter could be useful.

**Comment by Werkov** [comment]
Why doesn't start rate limiting resolve your need?

**Comment by martinvonwittich** [comment]
> Why doesn't start rate limiting resolve your need?

Configuring start rate limiting on the affected service would have made the problem with the faulty script much more obvious, you're right about that. But it would also require that each and every service would need to have a start rate limit set to catch misbehaving scripts, and it would also result in hard failure - a misbehaving script that restarts a service too often would result in the service failing completely until the interval has passed.

A persistent restart counter would work without any configuration beforehand, and it would allow monitoring without causing failures.

**Comment by Werkov** [comment]
> But it would also require that each and every service would need to have a start rate limit

FTR, you can add drop-in files in directories such as 
* `/etc/systemd/system/nameprefix-.service.d` or even
* `/etc/systemd/system/service.d` (see systemd.unit(5))

and configure service directives en masse.

The reason why I'm reluctant to the idea is that it requires a persistent storage (wrt systemd own process lifetime) -- and that kind of data is already captured in journald.

**Comment by poettering** [comment]
I am not convinced of this idea. If we'd count this then we could not GC units anymore (i.e. unload unreferenced, stopped units), because we'd then lose the counter.

**Comment by martinvonwittich** [comment]
> I am not convinced of this idea. If we'd count this then we could not GC units anymore (i.e. unload unreferenced, stopped units), because we'd then lose the counter.

Would it be possible to adapt the journal so that such persistent data could be stored alongside the logs?
